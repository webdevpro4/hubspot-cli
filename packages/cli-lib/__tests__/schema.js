const fs = require('fs-extra');
const path = require('path');
const chalk = require('chalk');
const { cleanSchema, writeSchemaToDisk, logSchemas } = require('../schema');
const { logger } = require('../logger');
const { getCwd } = require('../path');
const basic = require('./fixtures/schema/basic.json');
const full = require('./fixtures/schema/full.json');
const multiple = require('./fixtures/schema/multiple.json');

describe('cli-lib/schema', () => {
  const originalChalkLevel = chalk.level;
  beforeEach(() => {
    chalk.level = 0;
  });
  afterEach(() => {
    chalk.level = originalChalkLevel;
  });
  describe('cleanSchema()', () => {
    it('cleans a basic schema', () => {
      expect(cleanSchema(basic)).toMatchSnapshot();
    });

    it('cleans a full schema', () => {
      expect(cleanSchema(full)).toMatchSnapshot();
    });

    it('cleans multiple schema', () => {
      expect(multiple.map(cleanSchema)).toMatchSnapshot();
    });
  });

  describe('writeSchemaToDisk()', () => {
    it('writes schema to disk', () => {
      const spy = jest.spyOn(fs, 'outputFileSync');
      expect(fs.existsSync(path.resolve(getCwd(), `${basic.name}.json`))).toBe(
        false
      );
      writeSchemaToDisk(basic);
      expect(spy.mock.calls[0][1]).toMatchSnapshot();
    });
  });

  describe('logSchemas()', () => {
    it('logs schemas', () => {
      const spy = jest.spyOn(logger, 'log');
      logSchemas([basic]);
      expect(spy.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
