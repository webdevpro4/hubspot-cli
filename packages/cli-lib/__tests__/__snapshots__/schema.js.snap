// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli-lib/schema cleanSchema() cleans a basic schema 1`] = `
Object {
  "associatedObjects": null,
  "labels": Object {
    "plural": "Schemas",
    "singular": "Schema",
  },
  "name": "schema",
  "primaryDisplayProperty": "name",
  "properties": Array [
    Object {
      "description": "Name Field",
      "fieldType": "text",
      "label": "Name",
      "name": "name",
      "type": "string",
    },
  ],
  "requiredProperties": Array [],
  "searchableProperties": Array [],
}
`;

exports[`cli-lib/schema cleanSchema() cleans a full schema 1`] = `
Object {
  "associatedObjects": undefined,
  "labels": Object {
    "plural": "Listing Agents",
    "singular": "Listing Agent",
  },
  "name": "listing_agent",
  "primaryDisplayProperty": "agentfirstname",
  "properties": Array [
    Object {
      "description": undefined,
      "fieldType": "text",
      "label": "Agent First Name",
      "name": "agentfirstname",
      "type": "string",
    },
    Object {
      "description": undefined,
      "fieldType": "text",
      "label": "Agent Last Name",
      "name": "agentlastname",
      "type": "string",
    },
    Object {
      "description": undefined,
      "fieldType": "text",
      "label": "Photo",
      "name": "photo",
      "type": "string",
    },
  ],
  "requiredProperties": Array [
    "agentfirstname",
    "agentlastname",
  ],
  "searchableProperties": Array [
    "agentfirstname",
  ],
}
`;

exports[`cli-lib/schema cleanSchema() cleans multiple schema 1`] = `
Array [
  Object {
    "associatedObjects": undefined,
    "labels": Object {
      "plural": "Listing Agents",
      "singular": "Listing Agent",
    },
    "name": "listing_agent",
    "primaryDisplayProperty": "agentfirstname",
    "properties": Array [
      Object {
        "description": undefined,
        "fieldType": "text",
        "label": "Agent First Name",
        "name": "agentfirstname",
        "type": "string",
      },
      Object {
        "description": undefined,
        "fieldType": "text",
        "label": "Agent Last Name",
        "name": "agentlastname",
        "type": "string",
      },
      Object {
        "description": undefined,
        "fieldType": "text",
        "label": "Photo",
        "name": "photo",
        "type": "string",
      },
    ],
    "requiredProperties": Array [
      "agentfirstname",
      "agentlastname",
    ],
    "searchableProperties": Array [
      "agentfirstname",
    ],
  },
  Object {
    "associatedObjects": undefined,
    "labels": Object {
      "plural": "Property Listings",
      "singular": "Property Listing",
    },
    "name": "property-listing",
    "primaryDisplayProperty": "address",
    "properties": Array [
      Object {
        "description": undefined,
        "fieldType": "text",
        "label": "Address",
        "name": "address",
        "type": "string",
      },
      Object {
        "description": "How many bathrooms",
        "fieldType": "number",
        "label": "Bathrooms",
        "name": "bathrooms",
        "type": "number",
      },
      Object {
        "description": "How many bedrooms the listing has",
        "fieldType": "number",
        "label": "Bedrooms",
        "name": "bedrooms",
        "type": "number",
      },
      Object {
        "description": undefined,
        "fieldType": "textarea",
        "label": "Description",
        "name": "description",
        "type": "string",
      },
      Object {
        "description": "HTML title for object dynamic pages",
        "fieldType": "text",
        "label": "Dynamic Page Title",
        "name": "dynamic_page_title",
        "type": "string",
      },
      Object {
        "description": "The url slug for use in object dynamic pages",
        "fieldType": "text",
        "label": "Dynamic Page Slug",
        "name": "dynamic_page_uri",
        "type": "string",
      },
      Object {
        "description": undefined,
        "fieldType": "date",
        "label": "Open House Date",
        "name": "open_house_date",
        "type": "date",
      },
      Object {
        "description": "Price of property",
        "fieldType": "number",
        "label": "Price",
        "name": "price",
        "type": "number",
      },
      Object {
        "description": "Property Image",
        "fieldType": "text",
        "label": "Property Image",
        "name": "property_image",
        "type": "string",
      },
      Object {
        "description": "Square feet of the property ",
        "fieldType": "number",
        "label": "Square Feet",
        "name": "square_feet",
        "type": "number",
      },
    ],
    "requiredProperties": Array [
      "address",
    ],
    "searchableProperties": Array [
      "address",
    ],
  },
]
`;

exports[`cli-lib/schema logSchemas() logs schemas 1`] = `
"╔════════╤════════╤══════════════╗
║ Label  │ Name   │ objectTypeId ║
║ Schema │ schema │ 2-123        ║
╚════════╧════════╧══════════════╝
"
`;

exports[`cli-lib/schema writeSchemaToDisk() writes schema to disk 1`] = `
"{
  \\"name\\": \\"schema\\",
  \\"labels\\": { \\"singular\\": \\"Schema\\", \\"plural\\": \\"Schemas\\" },
  \\"requiredProperties\\": [],
  \\"searchableProperties\\": [],
  \\"primaryDisplayProperty\\": \\"name\\",
  \\"associatedObjects\\": null,
  \\"properties\\": [
    {
      \\"name\\": \\"name\\",
      \\"label\\": \\"Name\\",
      \\"type\\": \\"string\\",
      \\"fieldType\\": \\"text\\",
      \\"description\\": \\"Name Field\\"
    }
  ]
}
"
`;
